packages/@glimmer/validator/lib/tracking.ts

```diff
@@ -96,64 +115,133 @@ export function memo<T, U, V, W, X, Y, Z>(
-  let tag: Tag;
-  let snapshot: Revision;
-
-  let memoized = (...args: unknown[]): T => {
-    if (!tag || !validateTag(tag, snapshot)) {
-      beginTrackFrame();
-
-      try {
-        if (DEBUG) {
-          runInAutotrackingTransaction!(() => (lastValue = callback(...args)), debuggingContext);
-        } else {
-          lastValue = callback(...args);
-        }
-      } finally {
-        tag = endTrackFrame();
-        snapshot = valueForTag(tag);
-        consumeTag(tag);
-
-        // If the final tag is constant, then we know for sure that this
-        // memoized function can never change. There are times when this
-        // information is useful externally (i.e. in the append VM, it tells us
-        // whether or not to emit opcodes) so we expose it via a metadata weakmap.
-        if (tag === CONSTANT_TAG) {
-          IS_CONST_MAP.set(memoized, true);
-        } else if (DEBUG) {
-          // In DEBUG, set the value to false explicitly. This way we can throw
-          // if someone attempts to call `isConst(memoized)` before running
-          // `memoized()` at least once.
-          IS_CONST_MAP.set(memoized, false);
-        }
-      }
-    } else {
-      consumeTag(tag);
-    }
+  let cache = createCache(callback, debuggingContext);
+
+  let memoized = (...args: unknown[]) => getValue(cache, ...args);
+
+  (memoized as any)[CACHE_KEY] = cache;
+
+  return memoized;
+}
+
+export function isConstMemo(memoized: Function) {
+  let cache = (memoized as any)[CACHE_KEY] as Cache | undefined;
 
-    return lastValue!;
+  return cache === undefined ? false : isConst(cache);
+}
+
+//////////
+
+// public interface
+export interface Cache<T = unknown> {
+  [CACHE_KEY]: T;
+}
+
+const FN: unique symbol = symbol('FN');
+const LAST_VALUE: unique symbol = symbol('LAST_VALUE');
+const TAG: unique symbol = symbol('TAG');
+const SNAPSHOT: unique symbol = symbol('SNAPSHOT');
+const DEBUG_LABEL: unique symbol = symbol('DEBUG_LABEL');
+
+interface InternalCache<T = unknown> {
+  [FN]: (...args: unknown[]) => T;
+  [LAST_VALUE]: T | undefined;
+  [TAG]: Tag | undefined;
+  [SNAPSHOT]: Revision;
+  [DEBUG_LABEL]?: string | false;
+}
+
+export function createCache<T>(fn: () => T, debuggingLabel?: string | false): Cache<T> {
+  if (DEBUG && !(typeof fn === 'function')) {
+    throw new Error(
+      `createCache() must be passed a function as its first parameter. Called with: ${String(fn)}`
+    );
+  }
+
+  let cache: InternalCache<T> = {
+    [FN]: fn,
+    [LAST_VALUE]: undefined,
+    [TAG]: undefined,
+    [SNAPSHOT]: -1,
   };
 
   if (DEBUG) {
-    IS_CONST_MAP.set(memoized, undefined);
+    cache[DEBUG_LABEL] = debuggingLabel;
   }
 
-  return memoized;
+  return (cache as unknown) as Cache<T>;
 }
 
-export function isConstMemo(memoized: Function) {
-  if (DEBUG && IS_CONST_MAP.has(memoized) && IS_CONST_MAP.get(memoized) === undefined) {
+export function getValue<T>(cache: Cache<T>, ...args: unknown[]): T {
```

Comment:
Ideally we don't export types that have `...args` like this, can we tweak `memo` instead to avoid needing this there?